Ocaml unit tests

niss36 opened the pull request.

AnanyaHN27 committed:
For questions 2 and 3

AnanyaHN27 committed:
Rename backend/focs-2-3-tests.ml to backend/unit-tests/focs-2-3-tests.ml

niss36 committed:
Unit tests for 1997p1q6.

MatthewIreland commented on backend/unit-tests/1997p1q6-tests.ml:
```diff
@@ -0,0 +1,115 @@
+let suite =
+"1997p1q6TestSuite">:::
+[
+OUnitTest.TestLabel("testLastOnEmptyPath", OUnitTest.TestCase(OUnitTest.Short, testLastOnEmptyPath));
```
These unit test names are very undescriptive. Please could we say exactly *what* property we're testing?

MatthewIreland replied:
Each unit test should test exactly one property of the function, and it should be clear in what way the function is wrong if the unit test fails.

niss36 replied:
Right, this one is supposed to test whether the `last` function applied to a graph and the empty list returns the root. How about `testLastOnEmptyPathReturnsRootVertex`?

I'll see if I can break the other tests down into several more specific tests.

MatthewIreland replied:
This new name sounds great!

MatthewIreland commented on backend/unit-tests/1997p1q6-tests.ml:
```diff
@@ -0,0 +1,115 @@
+
+
+let suite =
+"1997p1q6TestSuite">:::
```
Should check mkgraph too -- we could just call the function with some hardcoded values that represent both finite and infinite graphs

niss36 replied:
Is that just to check it doesn't raise an exception? Because I'm not sure there's a way to test the returned value "represents" the graph it is given.

Edit: should we also test finite but cyclic graphs as a third case, or is it too redundant?

MatthewIreland replied:
Yep, I think it's just to check the type is correct and the function doesn't throw an exception when called.

niss36 replied:
I've added test cases for mkgraph (just need to add a cyclic graph case).

To check the type, would it be sensible to just write a function that takes datatype `g` as an input (with a type annotation) and call it on the result of `mkgraph`?

MatthewIreland replied:
Sound sgood to me!

MatthewIreland commented on backend/unit-tests/focs-2-3-tests.ml:
```diff
@@ -0,0 +1,148 @@
+(*helper functions*)
```
Can we split up this large file so we have one file per exercise, please?

MatthewIreland commented on backend/unit-tests/focs-2-3-tests.ml:
```diff
@@ -0,0 +1,148 @@
+(*helper functions*)
+
```
Where are the test suites defined (see examples provided)?

MatthewIreland commented on backend/unit-tests/focs-2-3-tests.ml:
```diff
@@ -0,0 +1,148 @@
+
+(*intersect*)
+
+let testOnEmptyLists = assert_equal [] (Submission.intersect [] []);;
```
These unit test names are very undescriptive. Please could we say exactly what property we're testing?

Each unit test should test exactly one property of the function, and it should be clear in what way the function is wrong if the unit test fails.

As an example, this would be `testIntersectionOfEmptyListsIsTheEmptyList`.

MatthewIreland commented on backend/unit-tests/focs-2-3-tests.ml:
```diff
@@ -0,0 +1,148 @@
+	let rndNum2 = (Random.int rndNum1) + 1 in
+	let rndList1 = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int rndNum1 ))) @ (intsTo (Random.int rndNum1 )) in
+	let rndList2 = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int rndNum2 ))) @ (intsTo (Random.int rndNum2 )) in
+	assert_equal false (findDupes (Submission.intersect rndList1 rndList2));;
```
Are there not assert_true and assert_false functions?

AnanyaHN27 replied:
Sorry, I was unaware that these functions even existed! This would make more sense.

MatthewIreland commented on backend/unit-tests/focs-2-3-tests.ml:
```diff
@@ -0,0 +1,148 @@
+	let rndNum2 = (Random.int rndNum1) + 1 in
+	let rndList1 = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int rndNum1 ))) @ (intsTo (Random.int rndNum1 )) in
+	let rndList2 = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int rndNum2 ))) @ (intsTo (Random.int rndNum2 )) in
+	assert_equal true (checkOrder (Submission.intersect rndList1 rndList 2) rndList1);;
```
Can we make this a bit simpler just by using two lists of ascending integers rather than random numbers?

MatthewIreland commented on backend/unit-tests/focs-2-3-tests.ml:
```diff
@@ -0,0 +1,148 @@
+	let rndList2 = (List.sort (fun _ _ -> ((Random.int 2) - 1)) (intsTo (Random.int rndNum2 ))) @ (intsTo (Random.int rndNum2 )) in
+	assert_equal false (findDupes (Submission.intersect rndList1 rndList2));;
+
+let testSubmissionOrder = 
```
Again, very undescriptive name. If this fails, how would I know what's wrong without spending hours trying to work out what this function does?

MatthewIreland commented on backend/unit-tests/focs-2-3-tests.ml:
```diff
@@ -0,0 +1,148 @@
+    in
+    aux [] list (List.length list);;
+
+(*intersect*)
```
Intersection tests don't currently check that the number of cons (::) operations performed does not exceed the number of elements in the result list. Can you override the cons function in the standard library and add a reference to count the number of times it's called?

MatthewIreland commented on backend/unit-tests/focs-2-3-tests.ml:
```diff
@@ -0,0 +1,148 @@
+
+let testOnEmptyList = assert_equal [] (Submission.exf (fun x-> x+1) []);;
+
+let testOnExample = assert_equal ([2;8] || [8;2]) (Submission.exf (fun x-> x+1) [9;3;2;2;8]);;
```
Can we test exf on some large, autogenerated lists please? Can use permutations function to make testing the output order easier.

MatthewIreland replied:
Also, can we put exf in a separate file, please?

AnanyaHN27 replied:
Would it be FoCS supervision 2 in one file and supervision 3 in another then?

MatthewIreland replied:
I think one file per question would be best. Currently there's one file for each of the exercises listed at http://svwork.ireland.cx/tasks.php and it would be good if the new tests could follow a similar convention.

niss36 committed:
Add tests for mkgraph and mkgraph2 (just checking no exception is raised); give more meaningful names to other tests; change the tests for last2: two tests check only the returned vertex (not count) for correctness, and one checks both returned vertex and count over a few random paths (each path is presented twice, to check the count value).

AnanyaHN27 committed:
Separate files for supervisions 2 and 3

Need to make palindrome tests still

AnanyaHN27 committed:
Delete focs-2-3-tests.ml

niss36 committed:
Add tests to check the types of mkgraph and mkgraph2; add tests for mkgraph and mkgraph2 on finite cyclic graphs.

AnanyaHN27 committed:
Update focs3-unit-tests.ml

AnanyaHN27 committed:
Add files via upload

AnanyaHN27 committed:
Need to figure out how to override :: operator

AnanyaHN27 committed:
Delete focs2-unit-tests

AnanyaHN27 committed:
Need to figure out how to override :: operator

niss36 committed:
Unit tests for 2015p1q2.

AnanyaHN27 committed:
Update focs2-unit-tests.ml

AnanyaHN27 committed:
Update focs3-unit-tests.ml

niss36 committed:
Fix test suite name

MatthewIreland commented on backend/unit-tests/2015p1q2-tests.ml:
```diff
@@ -0,0 +1,101 @@
+
+let testFlattenDoesntOmitItems test_ctxt =
+    let l = Submission.flatten (Submission.grid Submission.integers Submission.integers cantorPairingFunction)
+    and n = 50001 + Random.int 100000 in
```
Can we do this for more than one random n?

MatthewIreland commented on backend/unit-tests/2015p1q2-tests.ml:
```diff
@@ -0,0 +1,101 @@
+let testMapIntegersToStringsHasCorrectValues test_ctxt =
+    let l = Submission.map string_of_int Submission.integers
+    and n = 1 + Random.int 100000 in
+    assert_equal (string_of_int n) (nth n l)
```
Can we test more than one item? i.e. do this for more than one random n?

MatthewIreland commented on backend/unit-tests/2015p1q2-tests.ml:
```diff
@@ -0,0 +1,101 @@
+let testDiag test_ctxt =
+    let l = Submission.diag (Submission.map (fun n -> Submission.map (fun m -> n * m) Submission.integers) Submission.integers)
+    and n = 1 + Random.int 100000 in
+    assert_equal (n * n) (nth n l)
```
Is it difficult to do this for all diagonal elements?

MatthewIreland commented on backend/unit-tests/focs2-unit-tests.ml:
```diff
@@ -0,0 +1,162 @@
+	let rndList2 = (intsTo 300) in
+	assert_false (findDupes (Submission.intersect rndList1 rndList2));;
+
+(* how to properly override cons? *)
```
I guess comment can be removed now?

AnanyaHN27 committed:
Splitting FoCS3 into 2 distinct files

AnanyaHN27 committed:
Delete focs3-unit-tests.ml

AnanyaHN27 committed:
Update focs2-unit-tests.ml

niss36 committed:
Improve unit tests for 2015p1q2:
- Add a test case for the type of `diag`
- Change other type tests to check a string lazylist instead of int (seems more likely to be only int by mistake)
- Repeat 5 times the tests that check one randomly selected value (except for flatten)
- Use a more restricted random range to compensate for tests being repeated, and reduce running time variance.

niss36 committed:
Merge remote-tracking branch 'origin/ocaml-unit-tests' into ocaml-unit-tests

AnanyaHN27 committed:
Removed comments
